languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -0,0 +1,477 @@
+    name: str
+
+
+@dataclass
+class Role:
+    name: str
+    python_class: Any
+    permissions: List[Permission]
+    implied_roles: List[str]
+
+
+@dataclass
+class Resource:
+    name: str
+    python_class: Any
+    actions: Set[str]
+    roles: Set[str]
+
+
+def ensure_configured(func):
+    def wrapper(self, *args, **kwargs):
+        if not self.configured:
+            self._configure()
+        func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class OsoRoles:
+    def __init__(self, oso, sqlalchemy_base, user_model):
+        # @Q: Is this where I should create the models?
+        # Would this even work? Do these have to happen at a
+        # certain time or something to get created in the database?
+
+        self.session = None
+
+        user_pk_type = inspect(user_model).primary_key[0].type
+        user_pk_name = inspect(user_model).primary_key[0].name
+        user_table_name = user_model.__tablename__
+
+        # Tables for the management api to save data.
+        class UserRole(sqlalchemy_base):
+            __tablename__ = "user_roles"
+            id = Column(Integer, primary_key=True)
+            user_id = Column(
+                user_pk_type, ForeignKey(f"{user_table_name}.{user_pk_name}")
+            )
+            resource_type = Column(String)
+            resource_id = Column(String)  # Most things can turn into a string lol.
+            role = Column(String)
+
+        class Permission(sqlalchemy_base):
+            __tablename__ = "permissions"
+            id = Column(Integer, primary_key=True)
+            resource_type = Column(String)
+            name = Column(String)
+
+        class Role(sqlalchemy_base):
+            __tablename__ = "roles"
+            name = Column(String, primary_key=True)
+            resource_type = Column(String)
+
+        class RolePermission(sqlalchemy_base):
+            __tablename__ = "role_permissions"
+            id = Column(Integer, primary_key=True)
+            role = Column(String, ForeignKey("roles.name"))
+            permission_id = Column(Integer, ForeignKey("permissions.id"))
+
+        class RoleImplication(sqlalchemy_base):
+            __tablename__ = "role_implications"
+            id = Column(Integer, primary_key=True)
+            from_role = Column(String, ForeignKey("roles.name"))
+            to_role = Column(String, ForeignKey("roles.name"))
+
+        self.oso = oso
+        self.UserRole = UserRole
+        self.Permission = Permission
+        self.Role = Role
+        self.RolePermission = RolePermission
+        self.RoleImplication = RoleImplication
+
+        self.resources = {}
+        self.permissions = []
+        self.roles = {}
+        self.relationships = []
+
+        self.configured = False
+
+    def _configure(self):
+        # @TODO: ALLLLL of the validation needed for the role model.
+
+        # @TODO: Figure out where this session should really come from.
+        assert self.session
+
+        self.resources = {}
```

Comment:
skipped reading most of this

Reply:
One note is we may want to add some validation of the data in the policy itself with helpful error messages to make up for the fact that we don't have a way to check that the configuration data is the right structure.

Reply:
yeah there's a bunch of validation stuff that needs to go in here